File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution{
public:
    void solve(ListNode *&head){
        
        ListNode *prev=NULL, *tail=head;
        while(tail != NULL && tail->next != NULL){
            prev=tail;
            tail=tail->next;
        }
        prev->next=NULL;
        tail->next=head;
        head=tail;
    }

    ListNode *rotateRight(ListNode *head, int k){
        if(head == NULL)
            return NULL;
        int size=1;
        ListNode *temp=head;
        while(temp->next != NULL){
            size++;
            temp=temp->next;
        }
        k%=size;
        while(k--){
            solve(head);
        }

        return head;
       
    }
};

File: Medium/61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None or k == 0:
            return head
        size=1
        tail=head
        while tail.next:
            size+=1
            tail=tail.next

        tail.next=head
        k%=size
        for _ in range(size-k):
            tail=tail.next

        head=tail.next
        tail.next=None
        return head
        